File: 0314-binary-tree-vertical-order-traversal/0314-binary-tree-vertical-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        min_col, max_col = 0, 0
        queue = deque([(root, 0)])
        map = defaultdict(list)

        while queue:
            node, col = queue.popleft()
            min_col = min(min_col, col)
            max_col = max(max_col, col)
            map[col].append(node.val)
            if node.left:
                queue.append((node.left, col - 1))
            if node.right:
                queue.append((node.right, col + 1))

        res = []
        for col in range(min_col, max_col + 1):
            res.append(map[col])

        return res
        

File: 0314-binary-tree-vertical-order-traversal/0314-binary-tree-vertical-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalOrder(TreeNode* root) {
        if (root == nullptr)
            return {};

        int min_col = 0, max_col = 0;
        queue<pair<TreeNode*, int>> qu;
        unordered_map<int, vector<int>> map;
        qu.push({root, 0});
        vector<vector<int>> res;

        while (!qu.empty()) {
            auto p = qu.front();
            qu.pop();
            TreeNode* node = p.first;
            int col = p.second;
            map[col].push_back(node->val);
            if (node->left != nullptr) {
                qu.push({node->left, col - 1});
                min_col = min(min_col, col - 1);
            } 
            if (node->right != nullptr) {
                qu.push({node->right, col + 1});
                max_col = max(max_col, col + 1);
            }
        }

        for (int col = min_col; col <= max_col; col++) {
            res.push_back(map[col]);
        }

        return res;
    }
};